pypiserver/config.py

```diff
@@ -0,0 +1,599 @@
+  `RunConfig` for `run` options, and `UpdateConfig` for `update` options.
+- Ensure your config option is tested in `tests/test_config.py`.
+
+The `Config` class provides a `.from_args()` static method, which returns
+either a `RunConfig` or an `UpdateConfig`, depending on which subcommand
+is specified in the args.
+
+Legacy commandline arguments did not require a subcommand. This form is
+still supported, but deprecated. A warning is printing to stderr if
+the legacy commandline format is used.
+
+Command line arguments should be parsed as early as possible, using
+custom functions like the `auth_*` functions below if needed. For example,
+if an option were to take JSON as an argument, that JSON should be parsed
+into a dict by the argument parser.
+"""
+
+import argparse
+import contextlib
+import hashlib
+import io
+import pkg_resources
+import textwrap
+import sys
+import typing as t
+
+from pypiserver import __version__
+
+
+# Specify defaults here so that we can use them in tests &c. and not need
+# to update things in multiple places if a default changes.
+class DEFAULTS:
+    """Config defaults."""
+
+    AUTHENTICATE = ["update"]
+    FALLBACK_URL = "https://pypi.org/simple/"
+    HASH_ALGO = "md5"
+    INTERFACE = "0.0.0.0"
+    LOG_FRMT = "%(asctime)s|%(name)s|%(levelname)s|%(thread)d|%(message)s"
+    LOG_ERR_FRMT = "%(body)s: %(exception)s \n%(traceback)s"
+    LOG_REQ_FRMT = "%(bottle.request)s"
+    LOG_RES_FRMT = "%(status)s"
+    LOG_STREAM = sys.stdout
+    PACKAGE_DIRECTORIES = ["~/packages"]
+    PORT = 8080
+    SERVER_METHOD = "auto"
+
+
+def auth_arg(arg: str) -> t.List[str]:
+    """Parse the authentication argument."""
+    # Split on commas, remove duplicates, remove whitespace, ensure lowercase.
+    # Sort so that they'll have a consistent ordering.
+    items = sorted(list(set(i.strip().lower() for i in arg.split(","))))
+    # Throw for any invalid options
+    if any(i not in ("download", "list", "update", ".") for i in items):
+        raise ValueError(
+            "Invalid authentication option. Valid values are download, list, "
+            "and update, or . (for no authentication)."
+        )
+    # The "no authentication" option must be specified in isolation.
+    if "." in items and len(items) > 1:
+        raise ValueError(
+            "Invalid authentication options. `.` (no authentication) "
+            "must be specified alone."
+        )
+    return items
+
+
+def hash_algo_arg(arg: str) -> t.Callable:
+    """Parse a hash algorithm from the string."""
+    if arg not in hashlib.algorithms_available:
+        raise ValueError(
+            f"Hash algorithm {arg} is not available. Please select one "
+            f"of {hashlib.algorithms_available}"
+        )
+    return getattr(hashlib, arg)
+
+
+def html_file_arg(arg: t.Optional[str]) -> str:
+    """Parse the provided HTML file and return its contents."""
+    if arg is None or arg == "pypiserver/welcome.html":
```

Comment:
The only reason I didn't was because I'm currently using the formatter that automatically adds the defaults to the help text, and I didn't want it to look nasty. However, I think I need to switch to another formatter anyway so that I can get more granular formatting of some of the multiline helps with examples, so I'll make this change too